File: extensions/wasm/snitrafficpolicy/src/policy.rs
// Copyright 2026 The Kruise Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::error::Error;
use std::fmt::{Display, Formatter};

use prost::Message;

use crate::abi::{BoundPolicySnapshot, PolicySnapshotStatus, SniAction, SniTrafficPolicy};
use crate::matcher::matches_sni;
use crate::normalize::{normalize_pattern, normalize_sni};

pub const SNI_TRAFFIC_POLICY_TYPE_URL: &str =
    "type.googleapis.com/kruise.networking.extensions.v1.SniTrafficPolicy";

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct PolicyError(String);

impl Display for PolicyError {
    fn fmt(&self, formatter: &mut Formatter<'_>) -> std::fmt::Result {
        formatter.write_str(&self.0)
    }
}

impl Error for PolicyError {}

#[derive(Clone, Debug, Eq, PartialEq)]
struct Rule {
    patterns: Vec<String>,
    action: SniAction,
}

#[derive(Clone, Debug, Eq, PartialEq)]
struct Policy {
    resource_name: String,
    rules: Vec<Rule>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct PolicySnapshot {
    generation: u64,
    status: PolicySnapshotStatus,
    policies: Vec<Policy>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct PolicyDecision<'a> {
    pub resource_name: &'a str,
    pub rule_index: usize,
    pub action: SniAction,
}

impl PolicySnapshot {
    pub fn generation(&self) -> u64 {
        self.generation
    }

    pub fn status(&self) -> PolicySnapshotStatus {
        self.status
    }

    pub fn evaluate(&self, sni: &str) -> Result<Option<PolicyDecision<'_>>, PolicyError> {
        let normalized = normalize_sni(sni).map_err(|error| PolicyError(error.to_string()))?;
        for policy in &self.policies {
            for (rule_index, rule) in policy.rules.iter().enumerate() {
                if rule
                    .patterns
                    .iter()
                    .any(|pattern| matches_sni(pattern, &normalized))
                {
                    return Ok(Some(PolicyDecision {
                        resource_name: &policy.resource_name,
                        rule_index,
                        action: rule.action,
                    }));
                }
            }
        }
        Ok(None)
    }
}

impl TryFrom<BoundPolicySnapshot> for PolicySnapshot {
    type Error = PolicyError;

    fn try_from(snapshot: BoundPolicySnapshot) -> Result<Self, Self::Error> {
        let snapshot_status = PolicySnapshotStatus::try_from(snapshot.status).map_err(|_| {
            PolicyError(format!(
                "unknown policy snapshot status {}",
                snapshot.status
            ))
        })?;
        if snapshot_status == PolicySnapshotStatus::Unspecified {
            return Err(PolicyError("policy snapshot status is unspecified".into()));
        }

        if snapshot_status != PolicySnapshotStatus::Ready {
            return Ok(Self {
                generation: snapshot.generation,
                status: snapshot_status,
                policies: vec![],
            });
        }

        let mut sni_group = None;
        for group in snapshot.groups {
            if group.type_url != SNI_TRAFFIC_POLICY_TYPE_URL {
                continue;
            }
            if sni_group.replace(group).is_some() {
                return Err(PolicyError("duplicate SNI traffic policy group".into()));
            }
        }

        let Some(group) = sni_group else {
            return Ok(Self {
                generation: snapshot.generation,
                status: PolicySnapshotStatus::Ready,
                policies: vec![],
            });
        };
        let group_status = PolicySnapshotStatus::try_from(group.status).map_err(|_| {
            PolicyError(format!("unknown SNI policy group status {}", group.status))
        })?;
        if group_status == PolicySnapshotStatus::Unspecified {
            return Err(PolicyError("SNI policy group status is unspecified".into()));
        }
        if group_status != PolicySnapshotStatus::Ready {
            return Ok(Self {
                generation: snapshot.generation,
                status: group_status,
                policies: vec![],
            });
        }

        let mut policies = Vec::with_capacity(group.policies.len());
        for policy in group.policies {
            if policy.resource_name.is_empty() {
                return Err(PolicyError("policy resource name is empty".into()));
            }
            let resource =
                SniTrafficPolicy::decode(policy.resource.as_slice()).map_err(|error| {
                    PolicyError(format!(
                        "SNI policy {} cannot be decoded: {error}",
                        policy.resource_name
                    ))
                })?;
            let mut rules = Vec::with_capacity(resource.rules.len());
            for rule in resource.rules {
                let action = SniAction::try_from(rule.action)
                    .map_err(|_| PolicyError(format!("unknown SNI action {}", rule.action)))?;
                if action == SniAction::Unspecified {
                    return Err(PolicyError("SNI action is unspecified".into()));
                }
                let patterns = rule
                    .r#match
                    .ok_or_else(|| PolicyError("SNI match is missing".into()))?
                    .sni;
                if patterns.is_empty() {
                    return Err(PolicyError("SNI pattern list is empty".into()));
                }
                let patterns = patterns
                    .iter()
                    .map(|pattern| {
                        normalize_pattern(pattern).map_err(|error| PolicyError(error.to_string()))
                    })
                    .collect::<Result<Vec<_>, _>>()?;
                rules.push(Rule { patterns, action });
            }
            policies.push(Policy {
                resource_name: policy.resource_name,
                rules,
            });
        }
        Ok(Self {
            generation: snapshot.generation,
            status: PolicySnapshotStatus::Ready,
            policies,
        })
    }
}

File: extensions/epe/pkg/policy/trafficpolicy/policy.go
// Copyright 2026 The Kruise Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package trafficpolicy compiles TrafficPolicy egress rules into a small
// CONNECT-time authorizer. It deliberately evaluates only information present
// on the CONNECT request: the authenticated caller labels and destination
// host/port.
package trafficpolicy

import (
	"fmt"
	"net/netip"
	"strings"

	agentsv1alpha1 "github.com/openkruise/agents-api/agents/v1alpha1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/labels"
)

type peerKind uint8

const (
	peerUnsupported peerKind = iota
	peerCIDR
	peerFQDN
)

type peer struct {
	kind   peerKind
	prefix netip.Prefix
	fqdn   string
}

type port struct {
	protocol string
	start    int32
	end      int32
}

type rule struct {
	action agentsv1alpha1.RuleAction
	to     []peer
	ports  []port
}

// Policy is an immutable compiled TrafficPolicy or GlobalTrafficPolicy.
type Policy struct {
	Name              string
	Namespace         string
	Global            bool
	Priority          int32
	CreationTimestamp metav1.Time
	Version           string
	Selector          labels.Selector
	rules             []rule
	CompileError      string
}

// ResourceName implements krt.ResourceNamer. Scope prefixes prevent a global
// policy from colliding with a namespaced policy.
func (p Policy) ResourceName() string {
	if p.Global {
		return "global/" + p.Name
	}
	return "namespaced/" + p.Namespace + "/" + p.Name
}

func (p Policy) displayName() string {
	if p.Global {
		return "global/" + p.Name
	}
	return p.Namespace + "/" + p.Name
}

func compilePolicy(obj metav1.Object, spec *agentsv1alpha1.TrafficPolicySpec, global bool) (*Policy, error) {
	selector, err := metav1.LabelSelectorAsSelector(&spec.Selector)
	if err != nil {
		return nil, fmt.Errorf("selector: %w", err)
	}
	p := &Policy{
		Name:              obj.GetName(),
		Namespace:         obj.GetNamespace(),
		Global:            global,
		Priority:          spec.Priority,
		CreationTimestamp: obj.GetCreationTimestamp(),
		Version:           obj.GetResourceVersion(),
		Selector:          selector,
	}
	if spec.Egress == nil {
		return p, nil
	}
	p.rules = make([]rule, 0, len(spec.Egress.Rules))
	for i := range spec.Egress.Rules {
		compiled, err := compileRule(spec.Egress.Rules[i])
		if err != nil {
			return nil, fmt.Errorf("egress rule %d: %w", i, err)
		}
		p.rules = append(p.rules, compiled)
	}
	return p, nil
}

func invalidPolicy(obj metav1.Object, spec *agentsv1alpha1.TrafficPolicySpec, global bool, err error) *Policy {
	message := "invalid traffic policy"
	if err != nil {
		message = err.Error()
	}
	return &Policy{
		Name:              obj.GetName(),
		Namespace:         obj.GetNamespace(),
		Global:            global,
		Priority:          spec.Priority,
		CreationTimestamp: obj.GetCreationTimestamp(),
		Version:           obj.GetResourceVersion(),
		CompileError:      message,
	}
}

func compileRule(in agentsv1alpha1.TrafficPolicyRule) (rule, error) {
	if in.Action != agentsv1alpha1.RuleActionAllow && in.Action != agentsv1alpha1.RuleActionReject {
		return rule{}, fmt.Errorf("unsupported action %q", in.Action)
	}
	out := rule{action: in.Action}
	for _, raw := range in.To {
		compiled, err := compilePeer(raw)
		if err != nil {
			return rule{}, err
		}
		out.to = append(out.to, compiled)
	}
	for _, raw := range in.Ports {
		compiled, err := compilePort(raw)
		if err != nil {
			return rule{}, err
		}
		out.ports = append(out.ports, compiled)
	}
	return out, nil
}

func compilePeer(in agentsv1alpha1.TrafficPolicyPeer) (peer, error) {
	if in.CIDR != "" {
		prefix, err := netip.ParsePrefix(in.CIDR)
		if err != nil {
			addr, addrErr := netip.ParseAddr(in.CIDR)
			if addrErr != nil {
				return peer{}, fmt.Errorf("invalid CIDR %q: %w", in.CIDR, err)
			}
			prefix = netip.PrefixFrom(addr, addr.BitLen())
		}
		return peer{kind: peerCIDR, prefix: prefix.Masked()}, nil
	}
	if in.FQDN != "" {
		return peer{kind: peerFQDN, fqdn: normalizeHostname(in.FQDN)}, nil
	}
	// Service and Workload references need a second address-set watcher. Keep
	// them explicit and non-matching in this CONNECT-address PoC rather than
	// accidentally treating them as an unconstrained peer.
	return peer{kind: peerUnsupported}, nil
}

func compilePort(in agentsv1alpha1.TrafficPolicyPort) (port, error) {
	protocol := strings.ToUpper(in.Protocol)
	if protocol != "" && protocol != "TCP" && protocol != "UDP" && protocol != "ICMP" && protocol != "SCTP" {
		return port{}, fmt.Errorf("unsupported protocol %q", in.Protocol)
	}
	out := port{protocol: protocol}
	if in.Port != nil {
		out.start = *in.Port
		out.end = out.start
	}
	if in.EndPort != nil {
		if in.Port == nil || *in.EndPort < *in.Port {
			return port{}, fmt.Errorf("invalid port range")
		}
		out.end = *in.EndPort
	}
	return out, nil
}

func (r rule) matches(host string, destinationPort int32) bool {
	if len(r.ports) > 0 && !matchPorts(r.ports, destinationPort) {
		return false
	}
	if len(r.to) == 0 {
		return true
	}
	for _, candidate := range r.to {
		if candidate.matches(host) {
			return true
		}
	}
	return false
}

func matchPorts(ports []port, destinationPort int32) bool {
	for _, candidate := range ports {
		// CONNECT represents a TCP stream. A protocol-only TCP entry matches
		// every destination port, while UDP/ICMP/SCTP entries cannot match.
		if candidate.protocol != "" && candidate.protocol != "TCP" {
			continue
		}
		if candidate.start == 0 {
			return true
		}
		if destinationPort >= candidate.start && destinationPort <= candidate.end {
			return true
		}
	}
	return false
}

func (p peer) matches(host string) bool {
	switch p.kind {
	case peerCIDR:
		addr, err := netip.ParseAddr(strings.Trim(host, "[]"))
		return err == nil && p.prefix.Contains(addr)
	case peerFQDN:
		return matchHostname(p.fqdn, normalizeHostname(host))
	default:
		return false
	}
}

func normalizeHostname(host string) string {
	return strings.ToLower(strings.TrimSuffix(host, "."))
}

func matchHostname(pattern, host string) bool {
	if pattern == "*" {
		return true
	}
	if strings.HasPrefix(pattern, "*.") {
		suffix := pattern[1:]
		return len(host) > len(suffix) && strings.HasSuffix(host, suffix)
	}
	return pattern == host
}
